kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,118 @@ struct LogReplayScanner {
     seen: HashSet<FileActionKey>,
 }
 
-#[derive(Default)]
-struct AddRemoveVisitor {
-    adds: Vec<(Add, usize)>,
-    removes: Vec<Remove>,
-    selection_vector: Option<Vec<bool>>,
-    // whether or not we are visiting commit json (=true) or checkpoint (=false)
+/// A visitor that deduplicates a stream of add and remove actions into a stream of valid adds. Log
+/// replay visits actions newest-first, so once we've seen a file action for a given (path, dvId)
+/// pair, we should ignore all subsequent (older) actions for that same (path, dvId) pair. If the
+/// first action for a given file is a remove, then that file does not show up in the result at all.
+struct AddRemoveDedupVisitor<'seen> {
+    seen: &'seen mut HashSet<FileActionKey>,
+    selection_vector: Vec<bool>,
     is_log_batch: bool,
 }
 
-const ADD_FIELD_COUNT: usize = 15;
+impl<'seen> AddRemoveDedupVisitor<'seen> {
+    /// Checks if log replay already processed this logical file (in which case the current action
+    /// should be ignored). If not already seen, register it so we can recognize future duplicates.
+    fn already_seen(&mut self, key: FileActionKey) -> bool {
+        // Note: each (add.path + add.dv_unique_id()) pair has a
+        // unique Add + Remove pair in the log. For example:
+        // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
 
-impl AddRemoveVisitor {
-    fn new(selection_vector: Option<Vec<bool>>, is_log_batch: bool) -> Self {
-        AddRemoveVisitor {
-            selection_vector,
-            is_log_batch,
-            ..Default::default()
+        if self.seen.contains(&key) {
+            debug!(
```

Comment:
What's your criteria for putting in debug logging? I want to know when it might be useful to incorporate into my own code.

Reply:
I don't have any really concrete criteria... but if I added a `println!` for debugging that turned out to be especially useful, I'm a lot more likely to leave it there as a `debug!`?